### DISCUSSION TITLE
Is there any way to reflect ORM object custom define field by inspector?

### DISCUSSION THREAD

**Discussion by wangxin688** [sqlalchemy/sqlalchemy#10346] [Usage Questions] [answered]
I have the db model define as follow:
```python3
class HouseKeeping(Base):
    __tablename__ = "housekeeping"
   __multi_tenant__ = True
   __shared__ = True
   ...
async def inspect_table(table: str) -> dict:

    async with async_engine.connect() as conn:
        result = {"uq": [], "fk": {}}
        uq = await conn.run_sync(lambda sync_conn: inspect(sync_conn).get_unique_constraints(table_name=table))
        fk = await conn.run_sync(lambda sync_conn: inspect(sync_conn).get_foreign_keys(table_name=table))
```
I check the docs for Inspector here(https://docs.sqlalchemy.org/en/20/core/reflection.html#sqlalchemy.engine.reflection.Inspector.get_materialized_view_names)
, but I did not find a method to get custom define field like `__multi_tenant__` or `__shared__`. 

By the way, `Mapper` may not work for me, cause I don't find to a way get ORM object class by tablename

**Comment by CaselIT** [comment]
Hi,

> I did not find a method to get custom define field like __multi_tenant__ or __shared__.

I'm not sure what those fields are, there aren't an api by sqlalchemy, so I'm not sure how you would reflect them from the database.

I'm also unclear what tou are trying to do here.
The inspector you are mentioning uses the database to look at the tables, it doesn't use any of the defined models.
Could you clarify what you are looking for?

**Reply by wangxin688** [reply]
Sorry, I guess I was misunderstand the inpsector, it's reflect from database side. 
the `__shared__` and `__multi_tenant__` are costom fields define by myself. It's not taking effective to database side but only for class member variables for other usage. there's the basic code here.
```python
class Base(DeclarativeBase):  # dataclass_callable=dataclass
    """sqlalchemy orm base class, all orm object should inherit from this class, do not add any custom base

    Args:
        __multi_tenant__: default set as `True`, object will enable multi-tenant support automatically, orm cur
                          if set as `False`, object will disable multi-tenant support.\n
        __shared__: default set as `False`, object is multi_tenant, but need access common data defined in super-tenant.
                    common data is available for all tenants and maintained by super-tenant. Other tenant can create object
                    data only effective in self-tenant.\n
        __searchable_fields__: default set as `set()`, when searchable_fields is empty, object will not be searchable.\n
        __project_enable__: default set as `False`, use in buildout module.\n
        type_annotation_map: define column with `Mapped[uuid.UUId]` will mapped as `GUID` type
    """

    __multi_tenant__ = True
    __shared__ = False
    __project_enable__ = False
    type_annotation_map = {uuid.UUID: GUID}
    __searchable_fields__ = set()

    @declared_attr
    def tenant_id(cls) -> Mapped[uuid.UUID]:
        """auto generate tenant_id foreignkey when object is multi_tenant"""
        if not cls.__multi_tenant__:
            return None

        return mapped_column(GUID, ForeignKey("tenant.id", ondelete="CASCADE"), index=True, default=tenant_ctx.get)

    @declared_attr
    def is_shared(cls) -> Mapped[bool]:
        """this field is used for frontend control, if is_shared is True, cannot be edit or deleted by normal tenant.
        Common data should be created in `init_app.py` and set is_shared to `True` explicitly.
        when creating new instance from frontend by Tenant, is_shared should not be set in schema, default value will be set
        as `false`.
        """
        if not cls.__shared__:
            return None
        return mapped_column(Boolean, server_default=expression.false())
```
when other class inherit from `Base` and colunm will be autogenerate or not depends on it's magic variable settings.
So, I want to know if I can get this magic variables by tablename? 
in high level code, I use inspector to get forienkey  by `get_foreign_keys(table_name=table))`, it will return the forein table name, I need to the forien table's magic variables throught its tablename,  but I don't find a way to get ORM object class by tablename

**Reply by CaselIT** [reply, marked as answer]
You can consider using `Base.metadata.tables[table_name].foreign_keys` instead of `get_foreign_keys(table_name=table))`. This would avoid querying the db.

To get the orm class from the table name it not that that straight forward at the moment. This is a similar question to this issue https://github.com/sqlalchemy/sqlalchemy/issues/6080

You can probably look into `Base.registry.mappers`, loop on them looking for a `class_` with the table name you want. Something like
```py
def orm_by_table_name(table_name):
  for m in Base.registry.mappers:
    if getattr(m.class_, '__tablename__', None) == table_name:
      return m.class_
  return None
```

**Reply by wangxin688** [reply]
I guess that is what I wanted. thanks
